### PR TITLE
fix(cli): set `npm_config_user_agent` when running npm packages or tasks

### LINKED ISSUE #25342: Set npm_config_user_agent when running scripts
this is an env var that many scripts use to detect the package manager that ran the script. all major package managers including bun set this env var and it would be great if deno did too.

### PR DESCRIPTION
**Pull request by nathanwhit**
Fixes #25342.

Still not sure on the exact user agent to set (should it include `node`?).

After this PR, here's the state of running some `create-*` packages (just ones I could think of off the top of my head):

| package                  | prints/runs/suggests deno install | notes |
| ---------------- | ------------- | ------ |
| `create-next-app` | ❌ | falls back to npm, needs a PR ([code](https://github.com/vercel/next.js/blob/c32e2802097c03fd9f95b1dae228d6e0257569c0/packages/create-next-app/helpers/get-pkg-manager.ts#L3))
| `sv create` |  ❌ | uses `package-manager-detector`, needs a PR ([code](https://github.com/antfu-collective/package-manager-detector/tree/main))
| `create-qwik` | ✅  | runs `deno install` but suggests `deno start` which doesn't work (should be `deno task start` or `deno run start`)
| `create-astro` | ✅ | runs `deno install` but suggests `npm run dev` later in output, probably needs a PR
| `nuxi init` | ❌ | deno not an option in dialog, needs a PR ([code](https://github.com/nuxt/cli/blob/f04e2e894446f597da9d971b7eb03191d5a0da7e/src/commands/init.ts#L96-L102))
| `create-react-app` | ❌ | uses npm
| `ng new` (`@angular/cli`) | ❌ | uses npm
| `create-vite` | ✅ | suggests working deno commands 🎉 
| `create-solid` |  ❌ | suggests npm commands, needs PR

It's possible that fixing `package-manager-detector` or other packages might make some of these just work, but haven't looked too carefully at each

### CODE REVIEW

**Comment by birkskyum** [comment]
@nathanwhit , I had a look at adding deno to the solid-cli, but I found it to be incredibly tedious to work on. This is because I keep hitting errors with missing "node:" prefixes when testing this locally. All the deno node-compat that takes effect when running e.g. `deno run -A npm:create-solid` appear to be missing when building and executing the package locally `deno run index.ts` or `deno run index.mjs`, before it's published to npm.

**Comment by bartlomieju** [comment]
@birkskyum try running with `--unstable-sloppy-imports`. Then the `node:` prefixes shouldn't be necessary. I think the hint there is wrong, by not suggesting it.

**Comment by bartlomieju** [comment]
@birkskyum I was close :D it's actually `--unstable-bare-node-builtins`

**Comment by birkskyum** [comment]
That works better, is there a trick for allowing importing .json files (such as package.json) too?

`import { version } from "../package.json";`

Gives:

```sh
➜ deno run --unstable-bare-node-builtins ./packages/create-solid/src/index.ts     
Warning Resolving "fs" as "node:fs" at file:///Users/admin/repos/solid-cli/packages/commands/dist/handlers/new.mjs:1:485. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "util" as "node:util" at file:///Users/admin/repos/solid-cli/packages/utils/dist/index.mjs:5:158. If you want to use a built-in Node module, add a "node:" prefix.
error: Expected a JavaScript or TypeScript module, but identified a Json module. Consider importing Json modules with an import attribute with the type of "json".
  Specifier: file:///Users/admin/repos/solid-cli/packages/create-solid/package.json
    at file:///Users/admin/repos/solid-cli/packages/create-solid/src/index.ts:5:25
```

**Comment by bartlomieju** [comment]
No, I don't think we have a trick for that, it's strange that it allows to destructure JSON imports, I don't think it's supported in Node either. Shouldn't this code be somehow handled by a bundler first?

**Comment by birkskyum** [comment]
I can run the index.ts perfectly fine with `bun run`.

If I bundle it and run the dist/index.mjs with deno instead, I get this issue with `global`:

```sh
➜ deno run -A --unstable-bare-node-builtins ./packages/create-solid/dist/index.mjs
Warning Resolving "module" as "node:module" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:2:31. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "fs" as "node:fs" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:16:150. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "path" as "node:path" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:17:57. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "fs/promises" as "node:fs/promises" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:18:42. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "stream" as "node:stream" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:19:26. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "stream/promises" as "node:stream/promises" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:20:26. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "os" as "node:os" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:21:24. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "events" as "node:events" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:22:34. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "assert" as "node:assert" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:26:16. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "buffer" as "node:buffer" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:27:36. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "zlib" as "node:zlib" at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:28:16. If you want to use a built-in Node module, add a "node:" prefix.
error: Uncaught (in promise) ReferenceError: global is not defined
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:4663
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:459
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:5290
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:459
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:6869
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:459
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:8070
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:459
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:9542
    at file:///Users/admin/repos/solid-cli/packages/create-solid/dist/index.mjs:33:459
```

**Comment by birkskyum** [comment]
@bartlomieju , if I add `with {type: "json"}` to the import, and make it not-named like this:

`import packageJson from "../package.json" with { type: "json" };`

then deno can run the index.ts 🎉 

```sh
➜ deno run -A --unstable-bare-node-builtins ../packages/create-solid/src/index.ts
Warning Resolving "fs" as "node:fs" at file:///Users/admin/repos/solid-cli/packages/commands/dist/handlers/new.mjs:1:485. If you want to use a built-in Node module, add a "node:" prefix.
Warning Resolving "util" as "node:util" at file:///Users/admin/repos/solid-cli/packages/utils/dist/index.mjs:5:158. If you want to use a built-in Node module, add a "node:" prefix.
┌  
 Create-Solid v0.5.13
│
◇  Project Name
│  solid-start-created-with-deno
```

**Comment by bartlomieju** [comment]
@birkskyum no flag yet to enable `global`, but there's a PR that will add it: https://github.com/denoland/deno/pull/26617

**Comment by birkskyum** [comment]
Just realized it's a lot easier to work on adding support for deno in the cli tools _after_ this pr is merged.

File: cli/task_runner.rs [hunk 1]
```diff
@@ -204,7 +210,7 @@ impl ShellCommand for NpmCommand {
     mut context: ShellCommandContext,
   ) -> LocalBoxFuture<'static, ExecuteResult> {
     if context.args.first().map(|s| s.as_str()) == Some("run")
-      && context.args.len() > 2
+      && context.args.len() >= 2
```
Comment (bartlomieju) [inline cli/task_runner.rs:213]: Why is it now `>=`?

**Review by bartlomieju — APPROVED** [review]
LGTM 👍

**Comment by bartlomieju** [comment]
Opened a PR to `package-manager-detector`: https://github.com/antfu-collective/package-manager-detector/pull/25

File: cli/task_runner.rs (hunk 1 above)
Comment (nathanwhit) [reply cli/task_runner.rs:213]: Turned out this code wasn't triggering. `context.args` doesn't include the command name (`npm`), so the first arg is `run` and the second arg is the name of the script. So `npm run foo` was missing this branch.

**Comment by birkskyum** [comment]
Hmm, with canary deno 2.0.4+4774eab, the commit after this, I still see:

`console.log('agent:', process.env.npm_config_user_agent)` -> `agent: undefined`

**Comment by nathanwhit** [comment]
> Hmm, with canary deno 2.0.4+4774eab, the commit after this, I still see:
> 
> `console.log('agent:', process.env.npm_config_user_agent)` -> `agent: undefined`

It’s only set when running `npm:` entrypoints and running scripts currently. We could always set it, but not sure if that’s desirable. WDYT @bartlomieju?

**Comment by birkskyum** [comment]
@nathanwhit , ah that explains it

### So this doesn't work

`deno run --unstable-bare-node-builtins -A ./packages/create-solid/src/index.ts` -> agent: undefined


### And this does work:

configure package.json
```json
{
  "scripts": {
    "runwithagent": "deno run --unstable-bare-node-builtins -A ./packages/create-solid/src/index.ts",
  }
}
```

- `deno task runwithagent` -> `agent: deno/2.0.4 npm/? deno/2.0.4 macos aarch64`

**Comment by bartlomieju** [comment]
Hm, I'm not sure. Maybe it should also be set if there's a local `package.json` file?

**Pull request merged**